x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/AuthorizationEngine.java

```diff
@@ -152,13 +154,20 @@ void authorizeIndexAction(
      *                            alias or index
      * @param listener the listener to be notified of the authorization result
      */
+    // TODO: remove?
     void loadAuthorizedIndices(
         RequestInfo requestInfo,
         AuthorizationInfo authorizationInfo,
         Map<String, IndexAbstraction> indicesLookup,
         ActionListener<Set<String>> listener
     );
 
+    Predicate<IndexAbstraction> predicateForAuthorizedIndices(
+        RequestInfo requestInfo,
+        AuthorizationInfo authorizationInfo,
+        Metadata metadata
+    );
```

Comment:
Do you see an inherent advantage with using a predicate rather than a set?
(besides the delayed evalution, which I think should work for the set variant)

Reply:
Yes because it avoids building the Set for `Set#contains` calls, e.g.:

https://github.com/elastic/elasticsearch/blob/861dd90c8719742f004a882f09415d8fbab1fc12/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L85

https://github.com/elastic/elasticsearch/blob/861dd90c8719742f004a882f09415d8fbab1fc12/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L138

These are the places where the requested indices do not contain any wildcard. In this case, building the Set (of authorized indices) is not necessary. Testing directly against the predicate is most likely more efficient.

The predicate does not help when we do need iterate through the Set, e.g.:

https://github.com/elastic/elasticsearch/blob/861dd90c8719742f004a882f09415d8fbab1fc12/server/src/main/java/org/elasticsearch/cluster/metadata/IndexAbstractionResolver.java#L110

In this case, we have no choice but build the Set from the predicate. This step has the "double evaluation" issue over data stream names as you raised in the following comment. I'll address it over there.